Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp

```diff
@@ -1774,18 +1774,27 @@ void ResourceLoadStatisticsStore::grantStorageAccess(SubFrameDomain&& subFrameDo
             }
             ASSERT(subFrameStatus.first == AddedRecord::No);
 #if ASSERT_ENABLED
-            if (!NetworkStorageSession::canRequestStorageAccessForLoginOrCompatibilityPurposesWithoutPriorUserInteraction(subFrameDomain, topFrameDomain))
+            if (isEligibleForStorageAccessWithoutUserInteraction == IsEligibleForStorageAccessWithoutUserInteraction::No)
                 ASSERT(hasHadUserInteraction(subFrameDomain, OperatingDatesWindow::Long));
 #endif
             insertDomainRelationshipList(storageAccessUnderTopFrameDomainsQuery, HashSet<RegistrableDomain>({ topFrameDomain }), *subFrameStatus.second);
         }
 
-        grantStorageAccessInternal(WTFMove(subFrameDomain), WTFMove(topFrameDomain), frameID, pageID, promptWasShown, scope, WTFMove(completionHandler));
+        grantStorageAccessInternal(WTFMove(subFrameDomain), WTFMove(topFrameDomain), frameID, pageID, promptWasShown, scope, isEligibleForStorageAccessWithoutUserInteraction, WTFMove(completionHandler));
     };
 
-    RunLoop::main().dispatch([weakThis = WeakPtr { *this }, subFrameDomain = WTFMove(subFrameDomain).isolatedCopy(), topFrameDomain = WTFMove(topFrameDomain).isolatedCopy(), workQueue = m_workQueue, addGrant = WTFMove(addGrant), completionHandler = WTFMove(completionHandler)]() mutable {
-        auto additionalDomainGrants = NetworkStorageSession::storageAccessQuirkForDomainPair(subFrameDomain, topFrameDomain);
-        workQueue->dispatch([weakThis = WTFMove(weakThis), additionalDomainGrants = crossThreadCopy(WTFMove(additionalDomainGrants)), subFrameDomain = crossThreadCopy(WTFMove(subFrameDomain)), topFrameDomain = crossThreadCopy(WTFMove(topFrameDomain)), addGrant = WTFMove(addGrant), completionHandler = WTFMove(completionHandler)] () mutable {
+    RunLoop::main().dispatch([weakThis = WeakPtr { *this }, subFrameDomain = WTFMove(subFrameDomain).isolatedCopy(), topFrameDomain = WTFMove(topFrameDomain).isolatedCopy(), workQueue = m_workQueue, store = Ref { store() }, addGrant = WTFMove(addGrant), completionHandler = WTFMove(completionHandler)]() mutable {
+
+        std::optional<OrganizationStorageAccessPromptQuirk> additionalDomainGrants;
+        IsEligibleForStorageAccessWithoutUserInteraction isEligibleForStorageAccessWithoutUserInteraction { IsEligibleForStorageAccessWithoutUserInteraction::No };
+
+        if (auto* networkSession = store->networkSession()) {
+            if (auto* storageSession = networkSession->networkStorageSession()) {
+                additionalDomainGrants = storageSession->storageAccessQuirkForDomainPair(subFrameDomain, topFrameDomain);
+                isEligibleForStorageAccessWithoutUserInteraction = storageSession->canRequestStorageAccessForLoginOrCompatibilityPurposesWithoutPriorUserInteraction(subFrameDomain, topFrameDomain) ? IsEligibleForStorageAccessWithoutUserInteraction::Yes : IsEligibleForStorageAccessWithoutUserInteraction::No;
```

Comment:
isEligibleForStorageAccessWithoutUserInteraction is a somewhat inaccurate name.  It's not eligible for storage access, it's eligible to request storage access.  I'd rename it to canRequestStorageAccessWithoutUserInteraction or something like that with "request" in it.